File: csharp-playwright-framework/PlaywrightFramework/Pages/LoginPage.cs
using Microsoft.Playwright;

namespace PlaywrightFramework.Pages;

/// <summary>
/// Login Page Object Model
/// </summary>
public class LoginPage : BasePage
{
    // Locators
    private readonly ILocator _emailInput;
    private readonly ILocator _passwordInput;
    private readonly ILocator _loginButton;
    private readonly ILocator _errorMessage;
    private readonly ILocator _rememberMeCheckbox;
    private readonly ILocator _forgotPasswordLink;

    public LoginPage(IPage page) : base(page)
    {
        // Updated for saucedemo.com compatibility
        _emailInput = page.Locator("#user-name, #email, input[type='email'], input[name='email'], input[name='user-name']");
        _passwordInput = page.Locator("#password, input[type='password'], input[name='password']");
        _loginButton = page.Locator("#login-button, button[type='submit'], button:has-text('Login'), button:has-text('Sign in')");
        _errorMessage = page.Locator("[data-test='error'], .error-message, .alert-danger, [role='alert']");
        _rememberMeCheckbox = page.Locator("#remember-me, input[type='checkbox'][name='remember']");
        _forgotPasswordLink = page.Locator("a:has-text('Forgot'), a[href*='forgot']");
    }

    // Actions
    public async Task GotoAsync(string baseUrl)
    {
        // saucedemo.com login is on homepage, not /login
        await NavigateAsync(baseUrl);
        await WaitForPageLoadAsync();
    }

    public async Task LoginAsync(string email, string password)
    {
        await FillAsync(_emailInput, email);
        await FillAsync(_passwordInput, password);
        await ClickAsync(_loginButton);
    }

    public async Task LoginWithRememberMeAsync(string email, string password)
    {
        await FillAsync(_emailInput, email);
        await FillAsync(_passwordInput, password);
        await CheckAsync(_rememberMeCheckbox);
        await ClickAsync(_loginButton);
    }

    public async Task ClickForgotPasswordAsync()
    {
        await ClickAsync(_forgotPasswordLink);
    }

    // Assertions/Queries
    public async Task<string?> GetErrorMessageAsync()
    {
        return await GetTextAsync(_errorMessage);
    }

    public async Task<bool> IsErrorDisplayedAsync()
    {
        return await IsVisibleAsync(_errorMessage);
    }

    public async Task<bool> IsLoginButtonEnabledAsync()
    {
        return await IsEnabledAsync(_loginButton);
    }
}

File: python-playwright-framework/pages/login_page.py
"""
Login Page Object Model
Contains all elements and actions for the login page
"""

from playwright.sync_api import Page, Locator
from pages.base_page import BasePage


class LoginPage(BasePage):
    """Login page object model"""
    
    # Page URL
    LOGIN_URL = "/login"
    
    def __init__(self, page: Page):
        super().__init__(page)
        
        # Initialize locators
        self._email_input = page.locator('input[name="email"], input[type="email"], input#email')
        self._password_input = page.locator('input[name="password"], input[type="password"], input#password')
        self._login_button = page.locator('button:has-text("Login"), button[type="submit"]:has-text("Sign in"), button#login-button')
        self._signup_link = page.locator('a:has-text("Sign up"), a:has-text("Register"), a:has-text("Create account")')
        self._forgot_password_link = page.locator('a:has-text("Forgot"), a:has-text("Reset password")')
        self._error_message = page.locator('.error-message, .alert-danger, [role="alert"]')
        self._success_message = page.locator('.success-message, .alert-success')
        self._remember_me_checkbox = page.locator('input[type="checkbox"][name="remember"], input#remember-me')
        self._show_password_button = page.locator('button:has-text("Show"), [aria-label="Show password"]')
        self._login_form = page.locator('form')
    
    def goto(self) -> None:
        """Navigate to login page"""
        self.navigate(self.LOGIN_URL)
        self.wait_for_page_load()
    
    def enter_email(self, email: str) -> None:
        """Enter email address"""
        self.fill(self._email_input, email)
    
    def enter_password(self, password: str) -> None:
        """Enter password"""
        self.fill(self._password_input, password)
    
    def click_login(self) -> None:
        """Click login button"""
        self.click(self._login_button)
    
    def login(self, email: str, password: str) -> None:
        """Complete login flow"""
        self.enter_email(email)
        self.enter_password(password)
        self.click_login()
    
    def login_with_remember_me(self, email: str, password: str) -> None:
        """Login with remember me option"""
        self.enter_email(email)
        self.enter_password(password)
        self.click(self._remember_me_checkbox)
        self.click_login()
    
    def get_error_message(self) -> str:
        """Get error message text"""
        self.wait_for_element(self._error_message, 3000)
        return self.get_text(self._error_message)
    
    def is_error_displayed(self) -> bool:
        """Check if error message is displayed"""
        try:
            self.wait_for_element(self._error_message, 3000)
            return self.is_visible(self._error_message)
        except:
            return False
    
    def get_success_message(self) -> str:
        """Get success message text"""
        self.wait_for_element(self._success_message, 3000)
        return self.get_text(self._success_message)
    
    def is_success_displayed(self) -> bool:
        """Check if success message is displayed"""
        try:
            self.wait_for_element(self._success_message, 3000)
            return self.is_visible(self._success_message)
        except:
            return False
    
    def click_forgot_password(self) -> None:
        """Click forgot password link"""
        self.click(self._forgot_password_link)
    
    def click_signup(self) -> None:
        """Click signup/register link"""
        self.click(self._signup_link)
    
    def is_login_button_enabled(self) -> bool:
        """Check if login button is enabled"""
        return self.is_enabled(self._login_button)
    
    def is_login_form_displayed(self) -> bool:
        """Check if login form is displayed"""
        return self.is_visible(self._login_form)
    
    def toggle_password_visibility(self) -> None:
        """Toggle show/hide password"""
        self.click(self._show_password_button)
    
    def clear_form(self) -> None:
        """Clear login form"""
        self.clear(self._email_input)
        self.clear(self._password_input)
    
    def get_email_value(self) -> str:
        """Get email input value"""
        return self.get_attribute(self._email_input, 'value') or ''
    
    def is_remember_me_checked(self) -> bool:
        """Check if remember me is checked"""
        return self.is_checked(self._remember_me_checkbox)
